rpc/src/client/transport/router.rs

```diff
@@ -0,0 +1,182 @@
+//! Event routing for subscriptions.
+
+use crate::client::subscription::SubscriptionTx;
+use crate::event::Event;
+use std::collections::{HashMap, HashSet};
+use tracing::debug;
+
+/// Provides a mechanism for tracking [`Subscription`]s and routing [`Event`]s
+/// to those subscriptions.
+///
+/// [`Subscription`]: struct.Subscription.html
+/// [`Event`]: ./event/struct.Event.html
+#[derive(Debug)]
+pub struct SubscriptionRouter {
+    // A map of subscription queries to collections of subscription IDs and
+    // their result channels. Used for publishing events relating to a specific
+    // query.
+    subscriptions: HashMap<String, HashMap<String, SubscriptionTx>>,
+}
+
+impl SubscriptionRouter {
+    /// Publishes the given event to all of the subscriptions to which the
+    /// event is relevant. At present, it matches purely based on the query
+    /// associated with the event, and only queries that exactly match that of
+    /// the event's.
+    pub fn publish(&mut self, ev: &Event) -> PublishResult {
+        let subs_for_query = match self.subscriptions.get_mut(&ev.query) {
+            Some(s) => s,
+            None => return PublishResult::NoSubscribers,
+        };
+        // We assume here that any failure to publish an event is an indication
+        // that the receiver end of the channel has been dropped, which allows
+        // us to safely stop tracking the subscription.
+        let mut disconnected = HashSet::new();
+        for (id, event_tx) in subs_for_query {
+            if let Err(e) = event_tx.send(Ok(ev.clone())) {
+                disconnected.insert(id.clone());
+                debug!(
+                    "Automatically disconnecting subscription with ID {} for query \"{}\" due to failure to publish to it: {}",
+                    id, ev.query, e
+                );
+            }
+        }
+        // Obtain a mutable reference because the previous reference was
+        // consumed in the above for loop. We should panic if there are no
+        // longer any subscriptions for this query.
+        let subs_for_query = self.subscriptions.get_mut(&ev.query).unwrap();
+        for id in disconnected {
+            subs_for_query.remove(&id);
+        }
+        if subs_for_query.is_empty() {
+            PublishResult::AllDisconnected
+        } else {
+            PublishResult::Success
+        }
+    }
+
+    /// Immediately add a new subscription to the router without waiting for
+    /// confirmation.
+    pub fn add(&mut self, id: impl ToString, query: impl ToString, tx: SubscriptionTx) {
+        let query = query.to_string();
+        let subs_for_query = match self.subscriptions.get_mut(&query) {
+            Some(s) => s,
+            None => {
+                self.subscriptions.insert(query.clone(), HashMap::new());
+                self.subscriptions.get_mut(&query).unwrap()
+            }
+        };
+        subs_for_query.insert(id.to_string(), tx);
```

Comment:
what if the id already exists?

Reply:
It'd be overwritten.

My assumption here is that our UUIDs that we're automatically generating for subscriptions won't ever collide. I suppose that leaves a very small, albeit non-zero, non-deterministic probability of having a subscription dropped arbitrarily, but I'd imagine that networking failures leading to the same outcome would be far higher probability than that of a UUIDv4 collision.